### PR TITLE
MM-45556 Ability to mark a channel as unread

### PR DESCRIPTION
**Pull request by ballista01**
#### Summary

Implemented a feature to mark as unread ability on channels:

1. Alt + click on a channel in the LHS marks the last post as unread. ie the new messages line is above it
2. Add the “Mark as Unread” option on the … menu for a channel in the sidebar

#### Ticket Link

Github: https://github.com/mattermost/mattermost-server/issues/20600
JIRA: https://mattermost.atlassian.net/browse/MM-45556

#### Related Pull Requests

None

#### Screenshots

|  before  |  after  |
|----|----|
| LHS_channel_menu_before | LHS_channel_menu_after |

#### Release Note

```release-note
New feature: 
1. Alt + click on a channel in the LHS marks the last post as unread. ie the new messages line is above it
2. Add the “Mark as Unread” option on the … menu for a channel in the sidebar
```

### CODE REVIEW

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx
Comment (hmhealey) [inline components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:246]: @esethna The spec has Mark as Unread appearing after the Favourite and Mute options, but presumably it makes sense to have at the same spot as Mark as Read. At the same time, I feel like it makes sense to gave Favourite and Mute first.

Thoughts on which way to go with that?

File: components/sidebar/sidebar_channel/sidebar_channel_link/sidebar_channel_link.tsx
Comment (neallred) [inline components/sidebar/sidebar_channel/sidebar_channel_link/sidebar_channel_link.tsx:155]: This branch also should `event.preventDefault()`. Otherwise, since the backing HTML element is an anchor tag, alt+click tries to bring up the option to save the content at the link as an html document (at least on my Ubuntu Linux / Brave browser setup).

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx
Comment (esethna) [reply components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:246]: I'll let @rmonga71 make the call, you make a good point and think it makes sense at the top given it's a more commonly used action in a users workflow

**Review by neallred — APPROVED** [review]
LGTM pending preventing default on the alt click. Thanks, @Ballista01 !

**Review by Ballista01 — COMMENTED** [review]
@neallred It's done. My Firefox doesn't do anything other than mark the channel as unread but Vivaldi tries to save the link as document as you said. Thanks for your information.

**Review by hmhealey — CHANGES_REQUESTED** [review]
Thanks for the PR, @Ballista01! Unfortunately, I think we might need to make some changes to how we do this which probably requires a new API on the server.

The first time you open Mattermost, it only loads posts for the current channel and some other "preloaded" ones (currently just unread channels, but that might change so I don't want to rely on that too much), but that means that this won't work for unloaded channels. I see that there's a null check there, so it at least won't error out, but I think we want this to work for even channels that haven't been loaded.

I'm trying to decide if it would make more sense here to add a new API for marking a channel as unread (which would be nice since the server knows everything it needs here, but it would also require server changes) or if we should make this action fetch the posts from the server if it doesn't already have them (but then I'm worried if that might cause any weird issues by loading posts at a time we don't expect).

**Comment by rmonga71** [comment]
Looks good to me in the front-end but after a new channel was created I'm running into a bug where the actions menu is being blocked by the top of the window with no way to scroll on it. See screenshot below:
image

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx
Comment (rmonga71) [reply components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:246]: Thanks @esethna for looping me in. Agree with you both - I like Mark as Unread at the top so it's in the same spot as Mark as Read, and I'd expect it's the most common user action from that menu.

**Comment by rmonga71** [comment]
@esethna marking a threaded reply as unread should leave the channel read but caused my Threads view in the LHS to show as unread right? If yes then no issues - that is behaving as expected. Just wanted to confirm (that's how I'd expect it to work).

**Review by rmonga71 — DISMISSED** [review]
Looks good to me aside from the mentioned issue with the actions menu being cut off (https://github.com/mattermost/mattermost-webapp/pull/10755#issuecomment-1185063901)

**Comment by esethna** [comment]
> I think we want this to work for even channels that haven't been loaded

@hmhealey I'd agree with this, it should always work regardless of if the channel has been loaded in the background or not. 

-------

@rmonga71 yes that would be expected, this PR shouldn't change any behaviour related to marking threads unread. 

Sounds like there might be some hesitation regarding the approach from dev (see @hmhealey's comment above). Let's try to find an agreeable approach and hopefully @Ballista01 can help with any tweaks or further changes needed.

**Comment by Ballista01** [comment]
> > I think we want this to work for even channels that haven't been loaded
> 
> @hmhealey I'd agree with this, it should always work regardless of if the channel has been loaded in the background or not.
> 
> @rmonga71 yes that would be expected, this PR shouldn't change any behaviour related to marking threads unread.
> 
> Sounds like there might be some hesitation regarding the approach from dev (see @hmhealey's comment above). Let's try to find an agreeable approach and hopefully @Ballista01 can help with any tweaks or further changes needed.

I agree with @hmhealey that a new API call should be more reliable. If you decide to go that way, I can try to make another 2 PRs for the webapp and server later next week.

**Comment by Ballista01** [comment]
> Thanks @Ballista01! Looks good to me in the front-end but after a new channel was created I'm running into a bug where the actions menu is being blocked by the top of the window with no way to scroll on it. See screenshot below:
> 
> image

I think I haven't do anything to the LHS menu other than added an "Mark as Unread" entry. I checked with the current community server, and the problem also appears when the browser zoom level is too high (4k display, 150% by Windows * 250% by Firefox) that menu will be chopped either on the top or at the buttom. Resize the menu and add a scroll bar for it will help. However, since the problem also appears in the current release, maybe we should open another issue for it?

**Comment by rmonga71** [comment]
> I think I haven't do anything to the LHS menu other than added an "Mark as Unread" entry. I checked with the current community server, and the problem also appears when the browser zoom level is too high (4k display, 150% by Windows * 250% by Firefox) that menu will be chopped either on the top or at the buttom. Resize the menu and add a scroll bar for it will help. However, since the problem also appears in the current release, maybe we should open another issue for it? mattermost_LHS_menu_chopped

@Ballista01 you're right - thanks for calling this out. I've created a [separate issue](https://mattermost.atlassian.net/browse/MM-45772) for this, thanks! 

I'll defer to @hmhealey on the best approach to address the unloaded channel issue. @hmhealey - for my understanding are there any differences from an end-user perspective you'd expect between the 2 options you mention here: https://github.com/mattermost/mattermost-webapp/pull/10755#pullrequestreview-1038009110 ? What potential issues might be expected from fetching posts from non-preloaded channels? I'd assume it'd be impacts on performance?

**Comment by hmhealey** [comment]
Loading the posts when you click Mark Unread is probably the lowest effort overall, even if it would be slightly slower than a dedicated API because you're making two requests instead of one. I don't think that case is particularly likely though since I see this feature being more used if you forgot to mark a channel as unread after having just viewed it, so I don't think we need to go to the effort to optimize for it. If anything, we could also just make it optimistic so that the user wouldn't notice any lag anyway.

As for the potential issues, I don't think that should actually be a problem now that I've thought about it some more. We know that the channel is read, so we can just use the `getPosts` action from `mattermost-redux/actions/posts` to load the most recent posts in the channel, and we don't need to worry about the user being somewhere else in the channel.

Also, funnily enough, I'm also involved in another discussion about a similar change for marking unloaded threads as unread in the Threads view https://github.com/mattermost/mattermost-webapp/pull/10689

**Comment by ballista01** [comment]
Have been busy with things in life recently. I'll try to get it done when I have some time next week.

**Comment by rmonga71** [comment]
> Have been busy with things in life recently. I'll try to get it done when I have some time next week.

Sounds good, thanks @ballista01! Are you and @hmhealey aligned on approach to move forward?

**Comment by esethna** [comment]
@ballista01 how's this going? Let us know when you need a re-review?

**Comment by ballista01** [comment]
> @ballista01 how's this going? Let us know when you need a re-review?

I've added a step to load the posts for unloaded channels before marking the most recent post as unread.
Most of the logic is in ```markMostRecentPostInChannelAsUnread``` (```mattermost-redux/actions/channels```).

It goes like this, First, ```getState()```, then ```getMostRecentPostIdInChannel```

1. For loaded channels, we'll get a valid ```postId```, directly mark it as unread.
2. For unloaded channels, call ```getPosts``` from ```mattermost-redux/actions/posts``` to load the channel, refresh the state, and ```getMostRecentPostIdInChannel```. If we get a valid ```postId```, mark the corresponding post as unread. If we don't, the channel is empty, do nothing. 

I see the test server failed creation. Where can I see the logs?

**Comment by neallred** [comment]
For the failed server creation, I think what happened is that this test instance was old enough, that behind the scenes, it was deleted but that wasn't reflected here. If this happens in again (maybe in a future PR), you can try removing the server tag, waiting a minute or two, and re-adding it one more time. If it persistently fails, then you can reach out to a staff member for help. The logs mentioned are on a private, staff-only channel.

Recreated the server, and confirmed that I am now able to mark unloaded channels as unread (as long as there is at least one message in the channel, which makes sense to me!). Re-requesting review from @hmhealey

File: packages/mattermost-redux/src/actions/channels.ts
Comment (hmhealey) [inline packages/mattermost-redux/src/actions/channels.ts:1447]: We can actually remove this block since it'll this'll exit the if block and still hit the one below it. As written, this currently marks a channel as unread twice if the channel isn't loaded

**Review by hmhealey — CHANGES_REQUESTED** [review]

**Review by hmhealey — APPROVED** [review]

**Review by esethna — APPROVED** [review]

**Comment by esethna** [comment]
Sorry for the delay @ballista01 we've been at a company wide event, we just need to pass this through QA. Thanks for the contribution! cc// @jgilliam17

**Comment by jgilliam17** [comment]
Thanks @ballista01, I apologize for the delay.
Can you fix tests/checks so we can trigger e2es?
Looks good overall, one issue
- When using alt+click on the channel that is already unread it should open the channel - this is not working on the test server.
- Additionally, can you add a new Cypress test for this feature? Thank you.

**Review by nevyangelova — APPROVED** [review]

**Comment by esethna** [comment]
Just needs QA review cc// @jgilliam17

**Comment by ballista01** [comment]
> Thanks @ballista01, I apologize for the delay. Can you fix tests/checks so we can trigger e2es? Looks good overall, one issue
> 
>     * When using alt+click on the channel that is already unread it should open the channel - this is not working on the test server.
> 
>     * Additionally, can you add a new Cypress test for this feature? Thank you.

Thanks for pointing it out! I'll update the functional test after finishing my school works later this week.

**Comment by esethna** [comment]
Hi @ballista01, how's this going? Would love to get this in!

**Comment by ballista01** [comment]
> Hi @ballista01, how's this going? Would love to get this in!
>  

Just have some problem with '''make test-data'''. I run it after '''make run-server''' as stated in the developer setup but it seems to try to run another server and tells me the port 8065 is in use. And after some time the server terminates. I fail some E2E tests irrelevant to this issue probably because of lacking of the generated test data.

Is it a known issue?

**Comment by esethna** [comment]
cc// @cwarnermm on the docs stuff. Thanks for the find @ballista01!

**Comment by cwarnermm** [comment]
Thanks, @esethna! I've created [a card](https://community.mattermost.com/boards/team/rcgiyftm7jyrxnma1osd8zswby/bbusacfknofye5jgtmrufo75mnc/v5hakifb3i7nw5emhjtbrdua7uy/c9s3wbanrw7b3pbkz8am4om8bdr) for these updates.

**Review by jgilliam17 — COMMENTED** [review]
Thank you @ballista01 and apologies for delay
Tested, looks good to merge after few suggestions below.
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/257640-c55f9ad-onprem-ent-webapp-pr-10755-mm-ee-test:c55f9ad/mochawesome.html) - no PR related failures.
Please mention me when changes are made - thanks 🙂

**Comment by ballista01** [comment]
> Thank you @ballista01 and apologies for delay Tested, looks good to merge after few suggestions below. [E2E report](https://mattermost-cypress-report.s3.amazonaws.com/257640-c55f9ad-onprem-ent-webapp-pr-10755-mm-ee-test:c55f9ad/mochawesome.html) - no PR related failures. Please mention me when changes are made - thanks 🙂

I've committed them. Thanks for the help!

File: e2e/cypress/tests/integration/mark_as_unread/mark_as_unread_spec.js
Comment (jgilliam17) [inline e2e/cypress/tests/integration/mark_as_unread/mark_as_unread_spec.js:103]:
```suggestion
    it('MM-T5223 The latest post should appear unread after marking the channel as unread', () => {
```

Comment (ballista01) [reply e2e/cypress/tests/integration/mark_as_unread/mark_as_unread_spec.js]: commited

**Comment by jgilliam17** [comment]
@hmhealey can you please approve the last workflow and merge? thanks 🙂

**Review by jgilliam17 — APPROVED** [review]

**Comment by esethna** [comment]
@jgilliam17 or others, can you help merge if there's nothing blocking and reviews complete?

**Pull request merged**
